src/coprocessor/util.rs

```diff
@@ -17,33 +17,71 @@ use tipb::schema::ColumnInfo;
 use super::codec::datum::Datum;
 use super::codec::mysql;
 
-/// Get the smallest key which is larger than the key given.
-pub fn prefix_next(key: &[u8]) -> Vec<u8> {
-    let mut nk = key.to_vec();
-    if nk.is_empty() {
-        nk.push(0);
-        return nk;
+/// Convert the key to the smallest key which is larger than the key given.
+pub fn prefix_next(key: &mut Vec<u8>) {
+    if key.is_empty() {
+        key.push(0);
+        return;
     }
-    let mut i = nk.len() - 1;
+    let mut i = key.len() - 1;
+
+    // Add 1 to the last byte that is not 255, and set it's following bytes to 0.
     loop {
-        if nk[i] == 255 {
-            nk[i] = 0;
+        if key[i] == 255 {
+            key[i] = 0;
         } else {
-            nk[i] += 1;
-            return nk;
+            key[i] += 1;
+            return;
         }
         if i == 0 {
-            nk = key.to_vec();
-            nk.push(0);
-            return nk;
+            // All bytes are 255. Append a 0 to the key.
+            for byte in key.iter_mut() {
+                *byte = 255;
+            }
+            key.push(0);
+            return;
         }
         i -= 1;
     }
 }
 
+/// Check if `key1`'s prefix next equals to `key2`
+pub fn prefix_next_equals(key1: &[u8], key2: &[u8]) -> bool {
+    if key1.is_empty() {
+        return key2 == [0];
+    }
+    if key2.is_empty() {
+        return false;
+    }
+
+    let mut carry_pos = key1.len() - 1;
```

Comment:
I think you should compare length of key1 and key2 first.

Reply:
I prefer logic like bellowing, it is clear and more readable:

let len = key.len();
let next_len = next.len();
if len == next_len {
 ...
} else if len + 1 == next_le {
...
} else {
  false
}